### PR TITLE
Create eip-3778

### PR DESCRIPTION
**Pull request by CyberFork**
This is a standard for time-stamp provable NFTs, which extends [ERC-721](./eip-721.md).
Enables NFT to mapping and anchor things that have time features. 
And also the time-stamp proof of NFTs can be the condition of interaction with smart contracts.
It is called Time proof Non-fungible Token and is referred to as tpNFT in the subsequent descriptions.

When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

**Comment by alita-moore** [comment]
Hi! I'm a bot, and I wanted to automerge your PR, but couldn't because of the following issue(s):


	 -

**Comment by alita-moore** [comment]
sorry about all of this, I just added a new feature to the bot and was having issues. It's working now, though.

**Comment by CyberFork** [comment]
> sorry about all of this, I just added a new feature to the bot and was having issues -- closing and reopening runs the bot up to date with master. It's working now, though.

Doesn't matter.

Can I submit it like this? What else do I need to do?

**Comment by alita-moore** [comment]
> > sorry about all of this, I just added a new feature to the bot and was having issues -- closing and reopening runs the bot up to date with master. It's working now, though.
> 
> Doesn't matter.
> 
> Can I submit it like this? What else do I need to do?

it needs to be reviewed by an editor but they have already been mentioned so just gotta wait for @MicahZoltu to review :)

**Comment by MicahZoltu** [comment]
I don't do ERC reviews, so it'll have to wait for another editor.

File: EIPS/eip-3777.md [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+title: Time proof non-fungible Token
+author: Cyberforker(cyberforker@outlook.com)
+discussions-to: https://ethereum-magicians.org/t/a-time-proof-non-fungible-token/
+status: draft
+type: Standards Track
+category: ERC
+created: 2021-09-01
+requires: 165, 721
+---
+## Simple Summary
```
Comment (lightclient) [inline EIPS/eip-3777.md:12]:
```suggestion
## Simple Summary
```
This section is no longer valid, please use the `description` preamble field. (see the `eip-template.md` for more info)

File: EIPS/eip-3777.md [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+This is a standard for time-stamp provable NFTs, which extends [ERC-721](./eip-721.md).
+Enables NFT to mapping and anchor things that have time features. 
+And also the time-stamp proof of NFTs can be the condition of interaction with smart contracts.
+It is called Time proof Non-fungible Token and is referred to as tpNFT in the subsequent descriptions.
+## Abstract
+Based on the tokenURI parameter of ERC721, the plain text parameter and the encryption parameter are added, and the time proof is set for the update of these three parameters.
```
Comment (lightclient) [inline EIPS/eip-3777.md:18]:
```suggestion
Based on the `tokenURI` parameter of ERC-721, the plain text parameter and the encryption parameter are added, and the time proof is set for the update of these three parameters.

```

File: EIPS/eip-3777.md [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+- Prophecy: Make predictions and cast them into tpNFT with time proof ability, as proof of personal ability, or as a certificate for receiving bonuses.
+- Identity: Participating in a series of on-chain behaviors through tpNFT, including the prophecy mentioned in the previous point, revolves around tpNFT's timestamp proof, which has become a natural proof of identity.
+- Legacy: Binding assets to a tpNFT, others can receive part or all of the assets by submitting the preimage of encryption parameter(Hash or HashTree). We have another EIP to solve the front-running attacks.
+- Dynamic NFT: With tpNFT, we only need a little improvement then can get a brand new type of NFT which state on the chain will automatically change over time without consuming any gas fees. Such features can open up new areas for games, installation art, financial contracts, and so on.
+   
+Summarize: From a philosophical point of view, The NFT of the ERC721 standard is mainly used to anchor and map things related to space, while tpNFT is used for related to time.
```
Comment (lightclient) [inline EIPS/eip-3777.md:30]:
```suggestion
Summarize: From a philosophical point of view, The NFT of the ERC-721 standard is mainly used to anchor and map things related to space, while tpNFT is used for related to time.

```

File: EIPS/eip-3777.md [hunk 4]
```diff
@@ -0,0 +1,312 @@
+---
+eip: 3777
```
Comment (lightclient) [inline EIPS/eip-3777.md:2]:
```suggestion
eip: 3778
```

File: EIPS/eip-3777.md [hunk 5]
```diff
@@ -0,0 +1,312 @@
+---
+eip: 3777
+title: Time proof non-fungible Token
+author: Cyberforker(cyberforker@outlook.com)
+discussions-to: https://ethereum-magicians.org/t/a-time-proof-non-fungible-token/
+status: draft
```
Comment (lightclient) [inline EIPS/eip-3777.md:6]:
```suggestion
status: Draft

```

File: EIPS/eip-3777.md [hunk 6]
```diff
@@ -0,0 +1,312 @@
+---
+eip: 3777
+title: Time proof non-fungible Token
+author: Cyberforker(cyberforker@outlook.com)
```
Comment (lightclient) [inline EIPS/eip-3777.md:4]:
```suggestion
author: Cyberforker (@CyberFork)

```

File: EIPS/eip-3777.md [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+       3. In three update functions, if the value of the calldata and the parameter(URI or description or hashedproof) in the storage of tpNFT are both zero, the update will be skipped to save Gas and meet the needs of mint empty NFT.
+   
+**ERC-165 Interface**
+We chose Standard Interface Detection (ERC-165) to expose the interfaces that a tpNFT smart contract supports.
+
+## Backwards Compatibility
+ 1. Compatible with ERC165 and the implementation of the relevant interface and registration.
+ 2. Based on ERC721 and fully compatible, it is a superset of ERC721. If you only fill in the tokenURI and do not fill in the description and hash in the mint, and then set tpNFT to be constant, you will get a tpNFT that only has more tokenURI update timestamps than the standard NFT with tokenURI.
+## Test Cases
+### Mint
```
Comment (lightclient) [inline EIPS/eip-3777.md:276]: Please remove these empty sections.

File: EIPS/eip-3777.md [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    2. When updating the data of tpNFT, it is prohibited to directly modify the TimeProof struct in the _timeProofs mapping, must use the three internal methods _updateURI, _updateDescription, and _updateHashProof provided by the standard to update the data of tpNFT so that the updated content and the updated timestamp are in the same atom transaction and executed correctly.
+3. Keccak256 is selected as the function of Hash, it is widely used and its safety has been tested for a long time.
+    1. Anti-collision: Can not find two different data with the same Hash value.
+    2. Irreversible: Raw raw data cannot be obtained by reverse calculation of Hash value.
+    3. Leakage risk: We provide a simple function [genHashProof] to pass in the string and generate the hash value. It should be clarified that this function is for reference only. The preimage may be leaked due to RPC and other reasons when using this method of deploying the contract on the chain. Therefore for safety, if your preimage is very important, it is recommended that you copy the source code to the local offline or use other SDK to generates hash values ​​in a local offline environment.
+## References
```
Comment (lightclient) [inline EIPS/eip-3777.md:304]: This is not a valid EIP section, please remove.

File: EIPS/eip-3777.md [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+
+**Standards**
+
+1. [ERC-20](./eip-20.md) Token Standard.
+2. [ERC-165](./eip-165.md) Standard Interface Detection.
+3. [ERC-721](./eip-721.md) Non-Fungible Token Standard.
+## Copyright
+Copyright and related rights waived via CC0.
```
Comment (lightclient) [inline EIPS/eip-3777.md:312]:
```suggestion
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
```

**Review by lightclient — CHANGES_REQUESTED** [review]

File: EIPS/eip-3777.md (hunk 9 above)
Comment (CyberFork) [reply EIPS/eip-3777.md]: Thanks for guiding the revision, it has been corrected, please check again.

**Review by lightclient — CHANGES_REQUESTED** [review]
After reading your EIP a few times, I have a few issues with it:

1. I'm not sure why this needs to be an ERC vs. just an implementation detail of applications. Each of your examples in the motivation can already be done by simply posting data on-chain, and interpreting that data in a frontend. It's unclear how tokenizing provides a more standard interface than tx hash -> block number -> timestamp.
2. There are numerous grammatical errors. I try to be very lenient on the quality of ERC because I realize many people contribute with varying backgrounds and proficiency in English. Unfortunately I cannot merge this as-is. If you wish to pursue this EIP, I recommend editing the document rigorously - resolving grammatical errors and removing any first person / casual tone. Please use [EIP-1559](/EIPS/eip-1559.md), [EIP-2718](./EIPS/eip-2718.md), and [EIP-3529](./EIPS/eip-3529.md) as inspiration for well-written EIPs.

**Comment by CyberFork** [comment]
> After reading your EIP a few times, I have a few issues with it:
> 
> 1. I'm not sure why this needs to be an ERC vs. just an implementation detail of applications. Each of your examples in the motivation can already be done by simply posting data on-chain, and interpreting that data in a frontend. It's unclear how tokenizing provides a more standard interface than tx hash -> block number -> timestamp.
> 2. There are numerous grammatical errors. I try to be very lenient on the quality of ERC because I realize many people contribute with varying backgrounds and proficiency in English. Unfortunately I cannot merge this as-is. If you wish to pursue this EIP, I recommend editing the document rigorously - resolving grammatical errors and removing any first person / casual tone. Please use [EIP-1559](/ethereum/EIPs/blob/masterEIPS/eip-1559.md), [EIP-2718](/ethereum/EIPs/blob/master/EIPS/eip-2718.md), and [EIP-3529](/ethereum/EIPs/blob/master/EIPS/eip-3529.md) as inspiration for well-written EIPs.

1. Put the calldata in the transaction and through `transaction hash -> block number -> timestamp` to get timestamp is an excellent way to implement timestamp proofs. However, the problem is that any contracts cannot access it from the history transaction, can only get the current `block.timestamp`. We will highlight the need for tpNFT by putting the focus on examples about on-chain interactions.
2. We will carefully address this aspect based on feedback to make it more objective and accurate. For the other issues mentioned, we have already started to make changes and will contact someone with more expertise in English to help review the draft.

File: EIPS/eip-3778.md [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+category: ERC
+created: 2021-09-01
+requires: 165, 721
+---
+## Abstract
+This is a standard for timestamp provable NFTs, been called Time proof Non-fungible Token and is referred to as tpNFT in the subsequent descriptions, it enables NFT to mapping and anchor things that have time features, and also the timestamp proof of tpNFT can be the condition for smart contracts, this standard is extended from [ERC-721](./eip-721.md).
+
+Based on the "tokenURI" parameter of metadata of ERC-721 , the plain text parameter and the encryption parameter are added, and the timestamp proof is set for the update of these three parameters.
+With these improvements, we can provide new possibilities for NFT, enabling NFT to map and capture time-level value.
+Here are some brand new applications that can be achieved by tpNFT based on this extension. I believe you can find more and more interesting applications to bring more functions to the blockchain.
```
[the remark this answers is not in the document]
Comment (CyberFork) [reply EIPS/eip-3778.md]: Sorry about this content. We tried to write a similar sentence referencing "We considered a diverse universe of assets, and we know you would dream up more" in ERC-721. 
We will refer to EIP-1559, EIP-2718, and EIP-3529 as well-written EIPs.

**Review by lightclient — CHANGES_REQUESTED** [review]
This EIP still contains numerous grammatical errors. I am going to close this EIP as I've given this feedback previously, but it has not been addressed. If you would like to continue pursuing this EIP, I recommend opening a new PR with an updated draft that addresses these issues. Thanks!

**Pull request closed without merging**